Sources/libcoverage/coverage.c

```diff
@@ -100,36 +109,58 @@ void cov_shutdown(struct cov_context* context)
     shm_unlink(shm_key);
 }
 
+
 static int internal_evaluate(struct cov_context* context, uint8_t* virgin_bits, struct edge_set* new_edges)
 {
     uint64_t* current = (uint64_t*)context->shmem->edges;
     uint64_t* end = (uint64_t*)(context->shmem->edges + context->bitmap_size);
     uint64_t* virgin = (uint64_t*)virgin_bits;
-    
+
     new_edges->count = 0;
     new_edges->edges = NULL;
 
-    while (current < end) {
-        if (*current && unlikely(*current & *virgin)) {
-            // New edge(s) found!
+    // Only do more expensive tracking if corpus requires it
+    if(context->should_track_edges) {
+        while (current < end) {
             uint64_t index = ((uintptr_t)current - (uintptr_t)context->shmem->edges) * 8;
             for (uint64_t i = index; i < index + 64; i++) {
-                if (edge(context->shmem->edges, i) == 1 && edge(virgin_bits, i) == 1) {
-                    clear_edge(virgin_bits, i);
-                    new_edges->count += 1;
-                    new_edges->edges = realloc(new_edges->edges, new_edges->count * 4);
-                    new_edges->edges[new_edges->count - 1] = i;
+                if (edge(context->shmem->edges, i) == 1) {
+                    fflush(stdout);
+                    context->edge_count[i]++;
+                    fflush(stdout);
+                    if(edge(virgin_bits, i) == 1){
```

Comment:
Maybe it would be a bit nicer to always do the loop from lines 143 to 158 (for discovering whether there is new coverage), then only do this loop if `should_track_edges` but only process the edge counts. This will require two passes over the bitmap, but I suspect the first one will be significantly faster, so it shouldn't really matter. On the other hand, it would remove some code duplication and should be easier to follow. WDYT?

Reply:
I've implemented it as two loops, and there isn't a noticeable performance hit, so we should be good there!